### PR TITLE
Use generic labels in DCAs

### PR DESCRIPTION
**Pull request by leofeyer**
This PR implements the changes discussed in #509.

1. It is no loner necessary to add the `label` key in the DCA for (global) operations and fields. It will be set automatically if not given.

2. <del>It is possible to define a special `_element` label per table, which will be used to replace the word "element" with the label content, so e.g. "edit element ID 2" becomes "edit user group ID 2".</del>

3. The `alt` tags of the IMG elements are now the same as the `title` tags of the A elements, which fixes the accessibility of the buttons (see https://ux.stackexchange.com/a/92259).

@contao/developers Let me know if you like the concept, then I will complete the PR.

### TODO

* [x] XLIFF files
* [x] Other bundles

### CODE REVIEW

**Comment by ausi** [comment]
Will the `_element` label work for every language? Do we need a plural `_elements` too?

**Comment by Toflar** [comment]
In other languages there‘s not just singular and plural afaik. Or maybe there is but the rules are different. Maybe we can use doctrine/inflector? Or Symfony (not sure if English only).

**Comment by leofeyer** [comment]
> Will the `_element` label work for every language? Do we need a plural `_elements` too?

I am only using the `_element` label for singular nouns ("edit article", "show article settings", "delete article", etc.), so it should be fine. There have not been any plural nouns so far.

> Maybe we can use doctrine/inflector?

We are not inflecting anything. It is a simple replacement of the word "element" with whatever the content of `$GLOBALS['TL_LANG'][$table]['_element']` is.

File: core-bundle/src/Resources/contao/library/Contao/DcaLoader.php [hunk 1]
```diff
@@ -114,6 +116,99 @@ public function load($blnNoCache=false)
...
+		{
+			if (!isset($GLOBALS['TL_LANG'][$this->strTable][$k]) && isset($GLOBALS['TL_LANG']['DCA'][$k]))
+			{
+				if (\is_array($GLOBALS['TL_LANG']['DCA'][$k]))
+				{
+					$GLOBALS['TL_LANG'][$this->strTable][$k] = array_map
+					(
+						function ($str)
+						{
+							return str_replace(':element', $GLOBALS['TL_LANG'][$this->strTable]['_element'] ?? 'element', $str );
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/DcaLoader.php]: this will not work for translations

File: core-bundle/src/Resources/contao/library/Contao/DcaLoader.php [hunk 2]
```diff
@@ -114,6 +116,99 @@ public function load($blnNoCache=false)
...
+					continue;
+				}
+
+				if (isset($GLOBALS['TL_LANG'][$this->strTable][$k]))
+				{
+					$v['label'] = $GLOBALS['TL_LANG'][$this->strTable][$k];
+				}
+				elseif (isset($GLOBALS['TL_LANG']['DCA'][$k]))
+				{
+					$v['label'] = str_replace(':element', $GLOBALS['TL_LANG'][$this->strTable]['_element'] ?? 'element', $GLOBALS['TL_LANG']['DCA'][$k]);
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/DcaLoader.php]: This will only work for english

**Review by aschempp — CHANGES_REQUESTED** [review]
The `DcaLoader` does currently not load the language files (`DC_Table` does), so the translations wont work in all cases.

**Comment by leofeyer** [comment]
Unfortunately, `_element` does not work because of the definite article in German ("Details des Elements" vs "Details der Seite"). So I am going to revert this.

Nevertheless, I am not really happy with using "element" everywhere, because it decreases the accessibility IMHO. It is a lot clearer if a button says "delete member group ID %s" than if it says "delete element ID %s".

**Comment by leofeyer** [comment]
The PR is now ready to merge. Even though I decided against using "element" everywhere (see above), it was possible to remove a lot of lines in DCA and language files. Plus we are fixing the `alt` attributes of the buttons, so this is still an improvement.

**Comment by Toflar** [comment]
Plus we‘re making it easier for third party bundles. So this is actually a huge improvement :)

File: core-bundle/src/Resources/contao/library/Contao/DcaLoader.php [hunk 3]
```diff
@@ -114,6 +116,59 @@ public function load($blnNoCache=false)
...
+			foreach ($GLOBALS['TL_DCA'][$this->strTable]['list'][$key] as $k=>&$v)
+			{
+				if (isset($v['label']))
+				{
+					continue;
+				}
+
+				if (isset($GLOBALS['TL_LANG'][$this->strTable][$k]))
+				{
+					$v['label'] = $GLOBALS['TL_LANG'][$this->strTable][$k];
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/DcaLoader.php]: Is it on purpose that you didn’t use references to the `$GLOBALS['TL_LANG']` array here?

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/DcaLoader.php]: No. I have to fix this.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/DcaLoader.php]: Fixed in c8b6e628552a3e1520614ca743c93485f13a71aa.

**Review by ausi — APPROVED** [review]

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php [hunk 4]
```diff
@@ -447,8 +448,7 @@ public function showAll()
 		$return = $this->panel() . Message::generate() . '
 <div id="tl_buttons">'.((Input::get('act') == 'select') ? '
 <a href="'.$this->getReferer(true).'" class="header_back" title="'.StringUtil::specialchars($GLOBALS['TL_LANG']['MSC']['backBTTitle']).'" accesskey="b" onclick="Backend.getScrollOffset()">'.$GLOBALS['TL_LANG']['MSC']['backBT'].'</a> ' : '') . ((Input::get('act') != 'select' && !$blnClipboard && !$GLOBALS['TL_DCA'][$this->strTable]['config']['closed'] && !$GLOBALS['TL_DCA'][$this->strTable]['config']['notCreatable']) ? '
-<a href="'.$this->addToUrl($hrfNew).'" class="'.$clsNew.'" title="'.StringUtil::specialchars($ttlNew).'" accesskey="n" onclick="Backend.getScrollOffset()">'.$lblNew.'</a>
-<a href="'.$this->addToUrl('&amp;act=paste&amp;mode=move').'" class="header_new" title="'.StringUtil::specialchars($GLOBALS['TL_LANG'][$this->strTable]['move'][1]).'" onclick="Backend.getScrollOffset()">'.$GLOBALS['TL_LANG'][$this->strTable]['move'][0].'</a> ' : '') . ($blnClipboard ? '
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/drivers/DC_Folder.php:451]: why did you remove this operation?

Comment (leofeyer) [reply core-bundle/src/Resources/contao/drivers/DC_Folder.php:451]: It is no longer required. There are "upload files" link in each folder or you can upload files with drag and drop.

File: core-bundle/src/Resources/contao/languages/en/default.xlf [hunk 5]
```diff
@@ -1943,6 +1943,66 @@
...
+      <trans-unit id="DCA.all.1">
+        <source>Edit multiple elements at once</source>
+      </trans-unit>
+      <trans-unit id="DCA.toggleNodes.0">
+        <source>Toggle all</source>
+      </trans-unit>
+      <trans-unit id="DCA.toggleNodes.1">
+        <source>Expand or collapse all nodes</source>
+      </trans-unit>
+      <trans-unit id="DCA.new.0">
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/languages/en/default.xlf:1958]: Feels strange that `new` and a few other have to be an array, but most other operations do not?

File: core-bundle/src/Resources/contao/drivers/DC_Table.php [hunk 6]
```diff
@@ -4628,6 +4647,9 @@ protected function listView()
 			$firstOrderBy = $this->firstOrderBy;
 		}
 
+		// Check the default labels (see #509)
+		$labelNew = $GLOBALS['TL_LANG'][$this->strTable]['new'] ?? $GLOBALS['TL_LANG']['DCA']['new'];
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/drivers/DC_Table.php:4651]: If you gast this to an array, then the value would become the button if there is only one. No title then, but I guess thats fine?

File: core-bundle/src/Resources/contao/languages/en/tl_content.xlf [hunk 7]
```diff
@@ -639,75 +639,48 @@
         <source>Hide the author</source>
       </trans-unit>
       <trans-unit id="tl_content.new.0">
-        <source>New element</source>
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/languages/en/tl_content.xlf:642]: Can we not keep this as it was?

File: core-bundle/src/Resources/contao/languages/en/tl_content.xlf [hunk 8]
```diff
@@ -639,75 +639,48 @@
...
+        <source>Hide/unhide content element ID %s</source>
+      </trans-unit>
       <trans-unit id="tl_content.pasteafter.0">
         <source>Paste at the top</source>
       </trans-unit>
       <trans-unit id="tl_content.pasteafter.1">
         <source>Paste after content element ID %s</source>
       </trans-unit>
       <trans-unit id="tl_content.pastenew.0">
-        <source>Add new at the top</source>
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/languages/en/tl_content.xlf:690]: Changing labels that don't bring much benefit requires translators to translate everything all over again…

File: core-bundle/src/Resources/contao/languages/en/default.xlf (hunk 5 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/languages/en/default.xlf:1958]: Global operations have both a title and a link text and are therefore arrays. Regular operations have only a title (and an icon instead of a link text) and are therefore strings.

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php (hunk 4 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/drivers/DC_Folder.php:451]: uhm, so we also don't need the *New* button at the top of content elements (because there's a *new* button on each row)?

File: core-bundle/src/Resources/contao/languages/en/default.xlf [hunk 9]
```diff
@@ -969,13 +969,13 @@
...
       <trans-unit id="MSC.all.1">
-        <source>Edit multiple records at once</source>
+        <source>Edit multiple elements at once</source>
       </trans-unit>
       <trans-unit id="MSC.all_override.0">
         <source>Override multiple</source>
       </trans-unit>
       <trans-unit id="MSC.all_override.1">
-        <source>Override multiple records at once</source>
+        <source>Override multiple elements at once</source>
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/languages/en/default.xlf:978]: Are you intentionally change this? Just asking because you reverted the "element" stuff. "Elements" to me is for content elements, not database records in general.

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php (hunk 4 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/drivers/DC_Folder.php:451]: You cannot create new content elements by dragging them, can you? So yes. (BTW, there is still a "New" button in the file manager to create new folders.)

File: core-bundle/src/Resources/contao/languages/en/default.xlf (hunk 9 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/languages/en/default.xlf:978]: We have agreed on using "elements" by default.

https://github.com/contao/contao/blob/8befee845bd83a1fe10f317d1e0bebcc977b6766/core-bundle/src/Resources/contao/languages/en/default.xlf#L1958-L2005

**Comment by BugBuster1701** [comment]
Wenn die Übersetzungen fehlen für die generischen Labels und der Cache wird ungünstig aufgebaut, dann hat das ganz seltsame Auswirkungen. Siehe Slack.
https://contao.slack.com/archives/CK4J0KNDB/p1563661670012700
